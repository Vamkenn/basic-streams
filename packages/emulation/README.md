# Emulation for basic-streams

TODO: docs
