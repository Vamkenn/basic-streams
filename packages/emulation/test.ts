import {EventsList, emulate, t, v} from "./index"

expect.addSnapshotSerializer(EventsList.jestSerializer)

describe("emulate", () => {
  test("map", () => {
    expect(
      emulate(create => {
        return cb => create(t(10), v(1))(x => cb(x + 1))
      }),
    ).toMatchSnapshot()
  })

  test("filter", () => {
    expect(
      emulate(create => {
        return cb =>
          create(v(1), t(10), v(0), t(10), v(1))(x => {
            if (x > 0) cb(x)
          })
      }),
    ).toMatchSnapshot()
  })

  test("order of simultaneous values", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1))
        const b = create(t(10), v(2))
        return cb => {
          const disposeA = a(cb)
          const disposeB = b(cb)
          return () => {
            disposeA()
            disposeB()
          }
        }
      }),
    ).toMatchSnapshot()
  })

  test("unsubscribe", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1), t(10), v(0))
        const b = create(t(15), v(null))
        return cb => b(a(cb))
      }),
    ).toMatchSnapshot()
  })

  test("unsubscribe doesn't remove other values with same cb", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1), t(10), v(0))
        const b = create(t(10), v(1), t(10), v(0))
        const unsubscriber = create(t(15), v(null))
        return cb => {
          a(cb)
          unsubscriber(b(cb))
          return () => {}
        }
      }),
    ).toMatchSnapshot()
  })
})
