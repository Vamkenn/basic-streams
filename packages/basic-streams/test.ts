import {EventsList, emulate, t, v} from "@basic-streams/emulation"
import Stream from "./Stream"
import {
  empty,
  of,
  fromIterable,
  fromLoose,
  map,
  filter,
  chain,
  chainLatest,
  ap,
  map2,
  map3,
  combineArray,
  merge,
  scan,
  take,
  takeWhile,
  takeUntil,
  skip,
  skipWhile,
  skipDuplicates,
  startWith,
  multicast,
} from "./index"

expect.addSnapshotSerializer(EventsList.jestSerializer)

function noop() {}

function always(x: any): true {
  return true
}

function never(x: any): false {
  return false
}

function id<T>(x: T): T {
  return x
}

function append<T>(xs: T[], x: T): T[] {
  return xs.concat([x])
}

function inc(x: number): number {
  return x + 1
}

function dec(x: number): number {
  return x - 1
}

describe("empty", () => {
  test("subscriber is never called", () => {
    const cb = jest.fn()
    empty()(cb)()
    expect(cb.mock.calls).toMatchSnapshot()
  })
})

describe("of", () => {
  test("calls callback with the value", () => {
    const cb = jest.fn()
    of(1)(cb)
    expect(cb.mock.calls).toMatchSnapshot()
  })
})

describe("fromIterable", () => {
  test("works with array", () => {
    const cb = jest.fn()
    fromIterable([1, 2, 3])(cb)
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("works with generator", () => {
    const cb = jest.fn()
    fromIterable(
      (function*() {
        yield 1
        yield 2
        yield 3
      })(),
    )(cb)
    expect(cb.mock.calls).toMatchSnapshot()
  })
})

describe("fromLoose", () => {
  test("callback is called", () => {
    const result = emulate(create => {
      return fromLoose(create(t(10), v(1), t(10), v(2)))
    })
    expect(result).toMatchSnapshot()
  })

  test("disposer is called", () => {
    const disposer = jest.fn()
    fromLoose(() => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("fixes stream #3: It must return unsubscribe function (aka `disposer`)", () => {
    expect(typeof fromLoose(() => {})(noop)).toBe("function")
  })

  test("fixes stream #4: `cb` must be called with one argument", () => {
    const cb = jest.fn()
    fromLoose(inCb => {
      inCb(1, 2)
    })(cb)
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("fixes stream #5: `disposer` must always return `undefined`", () => {
    expect(fromLoose(() => () => 1)(noop)()).toBeUndefined()
  })

  test("fixes stream #6: After `disposer` was called, `cb` must not be called", () => {
    const cb = jest.fn()
    let inCb
    const unsub = fromLoose(_inCb => {
      inCb = _inCb
    })(cb)
    unsub()
    inCb(1)
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("fixes usage #1: `stream` must be called with one argument, `cb`", () => {
    const stream = jest.fn()
    ;(fromLoose(stream) as any)(noop, 1)
    expect(stream.mock.calls).toMatchSnapshot()
  })

  test("fixes usage #3: `cb` must always return `undefined`", () => {
    const subscriber = () => 1
    fromLoose(cb => {
      expect(cb(0)).toBeUndefined()
    })(subscriber)
  })

  test("fixes usage #4: `disposer` must be called with no arguments", () => {
    const disposer = jest.fn()
    ;(fromLoose(() => disposer)(noop) as any)(1, 2)
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("fixes usage #5: `disposer` must be called at most once", () => {
    const disposer = jest.fn()
    const disposer1 = fromLoose(() => disposer)(noop)
    disposer1()
    disposer1()
    expect(disposer.mock.calls).toMatchSnapshot()
  })
})

describe("map", () => {
  test("preserves disposer", () => {
    const disposer = jest.fn()
    map(id, cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("modifies values with provided fn", () => {
    const result = emulate(create => {
      return map(inc, create(t(10), v(1), t(10), v(2)))
    })
    expect(result).toMatchSnapshot()
  })
})

describe("filter", () => {
  test("preserves disposer", () => {
    const disposer = jest.fn()
    filter(always, cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("removes values", () => {
    const result = emulate(create => {
      return filter(x => x !== 1, create(t(10), v(1), t(10), v(2)))
    })
    expect(result).toMatchSnapshot()
  })
})

describe("chain", () => {
  test("preserves disposer of main sream", () => {
    const disposer = jest.fn()
    chain(x => of(x), cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("preserves disposer of spawned streams", () => {
    const disposer = jest.fn()
    chain(() => cb => disposer, of(null))(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("result stream contains values from spawned at correct time", () => {
    const result = emulate(create => {
      return chain(
        x => create(t(1), v(x * 2), t(1), v(x * 3)),
        create(t(10), v(1), t(10), v(2)),
      )
    })
    expect(result).toMatchSnapshot()
  })
})

describe("chainLatest", () => {
  test("preserves disposer of main sream", () => {
    const disposer = jest.fn()
    chainLatest(x => of(x), cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("preserves disposer of spawned streams", () => {
    const disposer = jest.fn()
    chainLatest(() => cb => disposer, of(null))(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("result stream contains values from spawned at correct time", () => {
    const result = emulate(create => {
      return chainLatest(
        x => create(t(1), v(x * 2), t(1), v(x * 3)),
        create(t(10), v(1), t(10), v(2)),
      )
    })
    expect(result).toMatchSnapshot()
  })

  test("doesn't contain values emited from old spawned streams", () => {
    const result = emulate(create => {
      return chainLatest(
        x => create(t(6), v(x * 2), t(6), v(x * 3)),
        create(t(10), v(1), t(10), v(2)),
      )
    })
    expect(result).toMatchSnapshot()
  })
})

describe("ap", () => {
  test("preserves disposers", () => {
    const disposer1 = jest.fn()
    const disposer2 = jest.fn()
    ap(() => disposer1, () => disposer2)(noop)()
    expect(disposer1.mock.calls).toMatchSnapshot()
    expect(disposer2.mock.calls).toMatchSnapshot()
  })

  test("updates result when inputs update", () => {
    const result = emulate(create => {
      return ap(
        create(t(6), v(inc), t(6), v(dec)),
        create(t(10), v(10), t(10), v(5)),
      )
    })
    expect(result).toMatchSnapshot()
  })
})

describe("map2", () => {
  test("preserves disposers", () => {
    const disposer1 = jest.fn()
    const disposer2 = jest.fn()
    map2((x, y) => [x, y], () => disposer1, () => disposer2)(noop)()
    expect(disposer1.mock.calls).toMatchSnapshot()
    expect(disposer2.mock.calls).toMatchSnapshot()
  })

  test("updates result when inputs update", () => {
    const result = emulate(create => {
      return map2(
        (x, y) => [x, y],
        create(t(6), v(1), t(6), v(2)),
        create(t(10), v(3), t(10), v(4)),
      )
    })
    expect(result).toMatchSnapshot()
  })
})

describe("map3", () => {
  test("preserves disposers", () => {
    const disposer1 = jest.fn()
    const disposer2 = jest.fn()
    const disposer3 = jest.fn()
    map3(
      (x, y, z) => [x, y, z],
      () => disposer1,
      () => disposer2,
      () => disposer3,
    )(noop)()
    expect(disposer1.mock.calls).toMatchSnapshot()
    expect(disposer2.mock.calls).toMatchSnapshot()
    expect(disposer3.mock.calls).toMatchSnapshot()
  })

  test("updates result when inputs update", () => {
    const result = emulate(create => {
      return map3(
        (x, y, z) => [x, y, z],
        create(t(7), v(1), t(7), v(2)),
        create(t(6), v(3), t(6), v(4)),
        create(t(10), v(5), t(10), v(6)),
      )
    })
    expect(result).toMatchSnapshot()
  })
})

describe("combineArray", () => {
  test("preserves disposers", () => {
    const disposer1 = jest.fn()
    const disposer2 = jest.fn()
    const disposer3 = jest.fn()
    combineArray([() => disposer1, () => disposer2, () => disposer3])(noop)()
    expect(disposer1.mock.calls).toMatchSnapshot()
    expect(disposer2.mock.calls).toMatchSnapshot()
    expect(disposer3.mock.calls).toMatchSnapshot()
  })

  test("result stream contains values from sources", () => {
    const result = emulate(create => {
      return combineArray([
        create(t(7), v(1), t(7), v(2)),
        create(t(6), v(3), t(6), v(4)),
        create(t(10), v(5), t(10), v(6)),
        create(v(7)),
      ])
    })
    expect(result).toMatchSnapshot()
  })

  test("works with one stream", () => {
    const result = emulate(create => {
      return combineArray([create(t(7), v(1), t(7), v(2))])
    })
    expect(result).toMatchSnapshot()
  })

  test("works with zero streams", () => {
    const result = emulate(create => {
      return combineArray([])
    })
    expect(result).toMatchSnapshot()
  })
})

describe("merge", () => {
  test("preserves disposers", () => {
    const disposer1 = jest.fn()
    const disposer2 = jest.fn()
    const disposer3 = jest.fn()
    merge([() => disposer1, () => disposer2, () => disposer3])(noop)()
    expect(disposer1.mock.calls).toMatchSnapshot()
    expect(disposer2.mock.calls).toMatchSnapshot()
    expect(disposer3.mock.calls).toMatchSnapshot()
  })

  test("result stream contains values from sources", () => {
    const result = emulate(create => {
      return merge([
        create(t(7), v(1), t(7), v(2)),
        create(t(6), v(3), t(6), v(4)),
        create(t(10), v(5), t(10), v(6)),
      ])
    })
    expect(result).toMatchSnapshot()
  })
})

describe("scan", () => {
  test("preserves disposer", () => {
    const disposer = jest.fn()
    scan(append, [], cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("contains correct values", () => {
    const result = emulate(create => {
      return scan(append, [], create(t(10), v(1), t(10), v(2)))
    })
    expect(result).toMatchSnapshot()
  })
})

describe("take", () => {
  test("calls disposer of source stream when we dispose result stream", () => {
    const disposer = jest.fn()
    take(1, () => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("doesn't call disposer twice", () => {
    const disposer = jest.fn()
    take(1, cb => {
      cb(1)
      return disposer
    })(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("takes first n and then calls disposer of source stream (async)", () => {
    const cb = jest.fn()
    let inCb = null
    take(2, _inCb => {
      inCb = _inCb
      return () => {
        inCb = null
      }
    })(cb)
    inCb(1)
    expect(typeof inCb).toBe("function")
    inCb(2)
    expect(inCb).toBe(null)
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("takes first n and then calls disposer of source stream (sync)", () => {
    const disposer = jest.fn()
    const cb = jest.fn()
    take(2, inCb => {
      inCb(1)
      inCb(2)
      inCb(3)
      return disposer
    })(cb)
    expect(disposer.mock.calls).toMatchSnapshot()
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("subscribes to source even if n is 0", () => {
    const stream = jest.fn(() => noop)
    take(0, stream)(noop)
    expect(stream.mock.calls).toMatchSnapshot()
  })

  test("contains correct values at correct time", () => {
    const result = emulate(create => {
      return take(1, create(t(10), v(1), t(10), v(2)))
    })
    expect(result).toMatchSnapshot()
  })

  test("if n is 0 returns empty stream", () => {
    const result = emulate(create => {
      return take(0, create(t(10), v(1), t(10), v(2)))
    })
    expect(result).toMatchSnapshot()
  })
})

describe("takeWhile", () => {
  test("calls disposer of source stream when we dispose result stream erlier", () => {
    const disposer = jest.fn()
    takeWhile(always, () => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("doesn't call disposer twice", () => {
    const disposer = jest.fn()
    takeWhile(never, cb => {
      cb(1)
      return disposer
    })(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("takes values that satisfy predicate until first value that don't then calls disposer (async)", () => {
    const cb = jest.fn()
    let inCb = null
    takeWhile(
      x => x < 3,
      _cb => {
        inCb = _cb
        return () => {
          inCb = null
        }
      },
    )(cb)
    inCb(1)
    inCb(2)
    expect(typeof inCb).toBe("function")
    inCb(3)
    expect(inCb).toBe(null)
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("takes values that satisfy predicate until first value that don't then calls disposer (sync)", () => {
    const disposer = jest.fn()
    const cb = jest.fn()
    takeWhile(
      x => x < 3,
      inCb => {
        inCb(1)
        inCb(2)
        inCb(3)
        inCb(4)
        return disposer
      },
    )(cb)
    expect(disposer.mock.calls).toMatchSnapshot()
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("return empty stream if predicate is () => false", () => {
    const result = emulate(create => {
      return takeWhile(never, create(t(10), v(1), t(10), v(2)))
    })
    expect(result).toMatchSnapshot()
  })

  test("contains correct values at correct time", () => {
    const result = emulate(create => {
      return takeWhile(
        x => x < 3,
        create(t(10), v(1), t(10), v(2), t(10), v(3), t(10), v(4)),
      )
    })
    expect(result).toMatchSnapshot()
  })
})

describe("takeUntil", () => {
  test("calls disposers when we dispose result stream erlier", () => {
    const disposer1 = jest.fn()
    const disposer2 = jest.fn()
    takeUntil(() => disposer1, () => disposer2)(noop)()
    expect(disposer1.mock.calls).toMatchSnapshot()
    expect(disposer2.mock.calls).toMatchSnapshot()
  })

  test("calls disposers after first value in controller", () => {
    const disposer1 = jest.fn()
    const disposer2 = jest.fn()
    takeUntil(cb => {
      cb(0)
      return disposer1
    }, () => disposer2)(noop)()
    expect(disposer1.mock.calls).toMatchSnapshot()
    expect(disposer2.mock.calls).toMatchSnapshot()
  })

  test("calls disposers properly (async value in controller)", () => {
    const disposer1 = jest.fn()
    const disposer2 = jest.fn()
    let cb = null
    const controller = _cb => {
      cb = _cb
      return disposer1
    }
    const disposer = takeUntil(controller, () => disposer2)(noop)
    cb && cb(0)
    disposer() // should't cause additional call of disposers
    expect(disposer1.mock.calls).toMatchSnapshot()
    expect(disposer2.mock.calls).toMatchSnapshot()
  })

  test("doesn't contain values after value from controller", () => {
    const result = emulate(create => {
      return takeUntil(
        create(t(25), v(1)),
        create(t(10), v(1), t(10), v(2), t(10), v(3)),
      )
    })
    expect(result).toMatchSnapshot()
  })

  test("doesn't contain values after sync value from controller", () => {
    const result = emulate(create => {
      return takeUntil(of(null), create(t(10), v(1), t(10), v(2), t(10), v(3)))
    })
    expect(result).toMatchSnapshot()
  })

  test("subscribes first to controller and then to source", () => {
    const cb = jest.fn()
    takeUntil(of(null), of(1))
    expect(cb.mock.calls).toMatchSnapshot()
  })
})

describe("skip", () => {
  test("preserves disposer", () => {
    const disposer = jest.fn()
    skip(0, cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("skips first N items", () => {
    const result = emulate(create => {
      return skip(2, create(t(10), v(1), t(10), v(2), t(10), v(3)))
    })
    expect(result).toMatchSnapshot()
  })

  test("returns equivalent stream if N is 0", () => {
    const result = emulate(create => {
      return skip(0, create(t(10), v(1), t(10), v(2), t(10), v(3)))
    })
    expect(result).toMatchSnapshot()
  })
})

describe("skipWhile", () => {
  test("preserves disposer", () => {
    const disposer = jest.fn()
    skipWhile(never, cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("skips first items that satisfy predicate", () => {
    const result = emulate(create => {
      return skipWhile(
        x => x < 3,
        create(t(10), v(1), t(10), v(2), t(10), v(3)),
      )
    })
    expect(result).toMatchSnapshot()
  })

  test("returns equivalent stream if predicate is () => false", () => {
    const result = emulate(create => {
      return skipWhile(never, create(t(10), v(1), t(10), v(2), t(10), v(3)))
    })
    expect(result).toMatchSnapshot()
  })

  test("returns empty stream if predicate is () => true", () => {
    const result = emulate(create => {
      return skipWhile(always, create(t(10), v(1), t(10), v(2), t(10), v(3)))
    })
    expect(result).toMatchSnapshot()
  })
})

describe("skipDuplicates", () => {
  function close(x: number, y: number): boolean {
    return Math.abs(x - y) < 1
  }

  test("preserves disposer", () => {
    const disposer = jest.fn()
    skipDuplicates(never, cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("doesn't apply comparator to the first element", () => {
    const comparator = jest.fn(close)
    skipDuplicates(comparator, of(1))(noop)
    expect(comparator.mock.calls).toMatchSnapshot()
  })

  test("removes duplicates", () => {
    const result = emulate(create => {
      return skipDuplicates(
        close,
        create(t(10), v(1), t(10), v(1.5), t(10), v(3)),
      )
    })
    expect(result).toMatchSnapshot()
  })
})

describe("startWith", () => {
  test("preserves disposer", () => {
    const disposer = jest.fn()
    startWith(0, cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("adds a value", () => {
    const result = emulate(create => {
      return startWith(0, create(t(10), v(1), t(10), v(2)))
    })
    expect(result).toMatchSnapshot()
  })
})

describe("multicast", () => {
  test("first subscribers gets sync events", () => {
    const cb = jest.fn()
    multicast(fromIterable([1, 2]))(cb)
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("second subscribers doesn't get sync events", () => {
    const cb = jest.fn()
    const stream = multicast(fromIterable([1, 2]))
    stream(noop)
    stream(cb)
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("first and second sunscribers get same async events", () => {
    const cb1 = jest.fn()
    const cb2 = jest.fn()
    let inCb = null
    const stream = multicast(_inCb => {
      inCb = _inCb
      return () => {
        inCb = null
      }
    })
    stream(cb1)
    stream(cb2)
    inCb(1)
    inCb(2)
    expect(cb1.mock.calls).toMatchSnapshot()
    expect(cb2.mock.calls).toMatchSnapshot()
  })

  test("subscriber doesn't get event after unsub() in response to that event", () => {
    const cb = jest.fn()
    let inCb = null
    let stream = multicast(_inCb => {
      inCb = _inCb
      return () => {
        inCb = null
      }
    })
    let unsub = null
    stream(x => {
      if (x === 2) {
        unsub && unsub()
      }
    })
    unsub = stream(cb)
    inCb(1)
    inCb(2)
    inCb(3)
    expect(cb.mock.calls).toMatchSnapshot()
  })

  test("preserves disposer (single subscriber)", () => {
    const disposer = jest.fn()
    multicast(cb => disposer)(noop)()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("preserves disposer (two subscribers)", () => {
    const disposer = jest.fn()
    const stream = multicast(cb => disposer)
    const unsub1 = stream(noop)
    const unsub2 = stream(noop)
    unsub1()
    unsub2()
    expect(disposer.mock.calls).toMatchSnapshot()
  })

  test("unsub doesn't remove another version of same subscriber", () => {
    const subscriber = jest.fn()
    let inCb = null
    const stream = multicast(_inCb => {
      inCb = _inCb
      return () => {
        inCb = null
      }
    })
    const unsub = stream(subscriber)
    stream(subscriber)
    unsub()
    unsub() // should be noop
    inCb(1)
    expect(subscriber.mock.calls).toMatchSnapshot()
  })
})
