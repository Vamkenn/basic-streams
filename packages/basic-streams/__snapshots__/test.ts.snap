// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ap preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`ap preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`ap updates result when inputs update 1`] = `
EventsList(
  10: 11
  12: 9
  20: 4
)
`;

exports[`chain preserves disposer of main sream 1`] = `
Array [
  Array [],
]
`;

exports[`chain preserves disposer of spawned streams 1`] = `
Array [
  Array [],
]
`;

exports[`chain result stream contains values from spawned at correct time 1`] = `
EventsList(
  11: 2
  12: 3
  21: 4
  22: 6
)
`;

exports[`chainLatest doesn't contain values emited from old spawned streams 1`] = `
EventsList(
  16: 2
  26: 4
  32: 6
)
`;

exports[`chainLatest preserves disposer of main sream 1`] = `
Array [
  Array [],
]
`;

exports[`chainLatest preserves disposer of spawned streams 1`] = `
Array [
  Array [],
]
`;

exports[`chainLatest result stream contains values from spawned at correct time 1`] = `
EventsList(
  11: 2
  12: 3
  21: 4
  22: 6
)
`;

exports[`combineArray preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`combineArray preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`combineArray preserves disposers 3`] = `
Array [
  Array [],
]
`;

exports[`combineArray result stream contains values from sources 1`] = `
EventsList(
  10: Array [
    1,
    3,
    5,
    7,
  ]
  12: Array [
    1,
    4,
    5,
    7,
  ]
  14: Array [
    2,
    4,
    5,
    7,
  ]
  20: Array [
    2,
    4,
    6,
    7,
  ]
)
`;

exports[`combineArray works with one stream 1`] = `
EventsList(
  7: Array [
    1,
  ]
  14: Array [
    2,
  ]
)
`;

exports[`combineArray works with zero streams 1`] = `
EventsList(
  0: Array []
)
`;

exports[`empty subscriber is never called 1`] = `Array []`;

exports[`filter preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`filter removes values 1`] = `
EventsList(
  20: 2
)
`;

exports[`fromIterable works with array 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
]
`;

exports[`fromIterable works with generator 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
]
`;

exports[`fromLoose callback is called 1`] = `
EventsList(
  10: 1
  20: 2
)
`;

exports[`fromLoose disposer is called 1`] = `
Array [
  Array [],
]
`;

exports[`fromLoose fixes stream #4: \`cb\` must be called with one argument 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`fromLoose fixes stream #6: After \`disposer\` was called, \`cb\` must not be called 1`] = `Array []`;

exports[`fromLoose fixes usage #1: \`stream\` must be called with one argument, \`cb\` 1`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`fromLoose fixes usage #4: \`disposer\` must be called with no arguments 1`] = `
Array [
  Array [],
]
`;

exports[`fromLoose fixes usage #5: \`disposer\` must be called at most once 1`] = `
Array [
  Array [],
]
`;

exports[`map modifies values with provided fn 1`] = `
EventsList(
  10: 2
  20: 3
)
`;

exports[`map preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`map2 preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`map2 preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`map2 updates result when inputs update 1`] = `
EventsList(
  10: Array [
    1,
    3,
  ]
  12: Array [
    2,
    3,
  ]
  20: Array [
    2,
    4,
  ]
)
`;

exports[`map3 preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`map3 preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`map3 preserves disposers 3`] = `
Array [
  Array [],
]
`;

exports[`map3 updates result when inputs update 1`] = `
EventsList(
  10: Array [
    1,
    3,
    5,
  ]
  12: Array [
    1,
    4,
    5,
  ]
  14: Array [
    2,
    4,
    5,
  ]
  20: Array [
    2,
    4,
    6,
  ]
)
`;

exports[`merge preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`merge preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`merge preserves disposers 3`] = `
Array [
  Array [],
]
`;

exports[`merge result stream contains values from sources 1`] = `
EventsList(
  6: 3
  7: 1
  10: 5
  12: 4
  14: 2
  20: 6
)
`;

exports[`multicast first and second sunscribers get same async events 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`multicast first and second sunscribers get same async events 2`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`multicast first subscribers gets sync events 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`multicast preserves disposer (single subscriber) 1`] = `
Array [
  Array [],
]
`;

exports[`multicast preserves disposer (two subscribers) 1`] = `
Array [
  Array [],
]
`;

exports[`multicast second subscribers doesn't get sync events 1`] = `Array []`;

exports[`multicast subscriber doesn't get event after unsub() in response to that event 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`multicast unsub doesn't remove another version of same subscriber 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`of calls callback with the value 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`scan contains correct values 1`] = `
EventsList(
  0: Array []
  10: Array [
    1,
  ]
  20: Array [
    1,
    2,
  ]
)
`;

exports[`scan preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`skip preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`skip returns equivalent stream if N is 0 1`] = `
EventsList(
  10: 1
  20: 2
  30: 3
)
`;

exports[`skip skips first N items 1`] = `
EventsList(
  30: 3
)
`;

exports[`skipDuplicates doesn't apply comparator to the first element 1`] = `Array []`;

exports[`skipDuplicates preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`skipDuplicates removes duplicates 1`] = `
EventsList(
  10: 1
  30: 3
)
`;

exports[`skipWhile preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`skipWhile returns empty stream if predicate is () => true 1`] = `EventsList()`;

exports[`skipWhile returns equivalent stream if predicate is () => false 1`] = `
EventsList(
  10: 1
  20: 2
  30: 3
)
`;

exports[`skipWhile skips first items that satisfy predicate 1`] = `
EventsList(
  30: 3
)
`;

exports[`startWith adds a value 1`] = `
EventsList(
  0: 0
  10: 1
  20: 2
)
`;

exports[`startWith preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`take calls disposer of source stream when we dispose result stream 1`] = `
Array [
  Array [],
]
`;

exports[`take contains correct values at correct time 1`] = `
EventsList(
  10: 1
)
`;

exports[`take doesn't call disposer twice 1`] = `
Array [
  Array [],
]
`;

exports[`take if n is 0 returns empty stream 1`] = `EventsList()`;

exports[`take subscribes to source even if n is 0 1`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`take takes first n and then calls disposer of source stream (async) 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`take takes first n and then calls disposer of source stream (sync) 1`] = `
Array [
  Array [],
]
`;

exports[`take takes first n and then calls disposer of source stream (sync) 2`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`takeUntil calls disposers after first value in controller 1`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers after first value in controller 2`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers properly (async value in controller) 1`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers properly (async value in controller) 2`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers when we dispose result stream erlier 1`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers when we dispose result stream erlier 2`] = `
Array [
  Array [],
]
`;

exports[`takeUntil doesn't contain values after sync value from controller 1`] = `EventsList()`;

exports[`takeUntil doesn't contain values after value from controller 1`] = `
EventsList(
  10: 1
  20: 2
)
`;

exports[`takeUntil subscribes first to controller and then to source 1`] = `Array []`;

exports[`takeWhile calls disposer of source stream when we dispose result stream erlier 1`] = `
Array [
  Array [],
]
`;

exports[`takeWhile contains correct values at correct time 1`] = `
EventsList(
  10: 1
  20: 2
)
`;

exports[`takeWhile doesn't call disposer twice 1`] = `
Array [
  Array [],
]
`;

exports[`takeWhile return empty stream if predicate is () => false 1`] = `EventsList()`;

exports[`takeWhile takes values that satisfy predicate until first value that don't then calls disposer (async) 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`takeWhile takes values that satisfy predicate until first value that don't then calls disposer (sync) 1`] = `
Array [
  Array [],
]
`;

exports[`takeWhile takes values that satisfy predicate until first value that don't then calls disposer (sync) 2`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;
